lib/ansible/plugins/inventory/__init__.py

```diff
@@ -288,28 +288,42 @@ def _add_host_to_composed_groups(self, groups, variables, host, strict=False):
+
         if keys and isinstance(keys, list):
+            groups = []
             for keyed in keys:
                 if keyed and isinstance(keyed, dict):
-                    prefix = keyed.get('prefix', '')
-                    key = keyed.get('key')
-                    if key is not None:
-                        try:
-                            groups = to_safe_group_name('%s_%s' % (prefix, self._compose(key, variables)))
-                        except Exception as e:
-                            if strict:
-                                raise AnsibleOptionsError("Could not generate group on %s: %s" % (key, to_native(e)))
+
+                    try:
+                        key = self._compose(keyed.get('key'), variables)
+                    except Exception as e:
+                        msg = "Could not generate group on %s: %s" % (keyed.get('key'), to_native(e))
+                        if strict:
+                            raise AnsibleOptionsError(msg)
+                        else:
+                            self.display.warning(msg)
                             continue
-                        if isinstance(groups, string_types):
-                            groups = [groups]
-                        if isinstance(groups, list):
-                            for group_name in groups:
-                                if group_name not in self.inventory.groups:
-                                    self.inventory.add_group(group_name)
-                                self.inventory.add_child(group_name, host)
+
+                    if key is not None:
+                        prefix = keyed.get('prefix', '')
+                        sep = keyed.get('separator', '_')
+
+                        if isinstance(key, string_types):
+                            groups.append(to_safe_group_name('%s%s%s' % (prefix, key, sep)))
+                        elif isinstance(key, list):
+                            for name in key:
+                                groups.append(to_safe_group_name('%s%s%s' % (prefix, sep, name)))
+                        elif isinstance(key, MutableMapping):
+                            for (gname, gval) in key.items():
+                                groups.append(to_safe_group_name('%s%s%s' % ('%s%s%s' % (gname, prefix, gval), prefix, sep)))
                         else:
                             raise AnsibleOptionsError("Invalid group name format, expected string or list of strings, got: %s" % type(groups))
                     else:
                         raise AnsibleOptionsError("No key supplied, invalid entry")
                 else:
                     raise AnsibleOptionsError("Invalid keyed group entry, it must be a dictionary: %s " % keyed)
+
+            # now actually add any groups
+            for group_name in groups:
+                if group_name:
+                    self.inventory.add_child(group_name, host)
```

Comment:
Need to ensure that the group actually exists - currently when using this, the inventory plugin fails here because the previous code containing 

```
                                if group_name not in self.inventory.groups:
                                    self.inventory.add_group(group_name)
```

has not been added here.